### PR TITLE
add docs about insertion deduplication on retries

### PR DESCRIPTION
**Pull request by CheSema**

### CODE REVIEW

File: docs/en/guides/developer/insertion-deduplication-on-retries.md [hunk 1]
```diff
@@ -0,0 +1,390 @@
+---
+slug: /en/guides/developer/insertion-deduplication-on-retries
+sidebar_label: Insertion deduplication on retries
+sidebar_position: 0
+description: On retries inderted data is deduplicated.
+---
+
+# What is deduplication on insertion retries?
+Some times the result of insert operation is unsertan. For example the case when user gets timeout error. The inserted data could be actually inserted to the destination or not.
```
Comment (kssenii) [inline docs/en/guides/developer/insertion-deduplication-on-retries.md:9]:
```suggestion
Some times the result of insert operation is uncertain. For example the case when user gets timeout error. The inserted data could actually be inserted to the destination or not inserted.
```

File: docs/en/guides/developer/insertion-deduplication-on-retries.md [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+Such failed operations should be retried by the user. When an user retries that operations Clickhouse tries to determine if the inserted data has been succesfuly inserted already or not. If that inserted data is marked as duplication than Clickhouse would not insert it to the destination table and user would recieve successful status of operation as if the data has been inserted.
```
Comment (kssenii) [inline docs/en/guides/developer/insertion-deduplication-on-retries.md:10]:
```suggestion
Such failed operations should be retried by the user. When a user retries that operations Clickhouse tries to determine if the inserted data has been succesfuly inserted already or not. If that inserted data is marked as duplication then Clickhouse would not insert it to the destination table and user would recieve successful status of operation as if the data has been inserted.
```

File: docs/en/guides/developer/insertion-deduplication-on-retries.md [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+
+# How enable or disable deduplication of insertion?
+Only the enginges from the family of engines `*MergeTree` support the deduplication on insertion.
+For engines from `*ReplicatedMergeTree` engine family deduplication is controlled by the settings: `replicated_deduplication_window` and `replicated_deduplication_window_seconds`.
+For non-replicated engine family `*MergeTree` deduplication is controlled by the settings: `non_replicated_deduplication_window`.
+That settings determine parameters of deduplication log for the table. Deduplication log stores finite count of `block_id`'s. That set of `block_id`'s determine how deduplication works.
```
Comment (kssenii) [inline docs/en/guides/developer/insertion-deduplication-on-retries.md:16]:
```suggestion
These settings determine parameters of deduplication log for the table. Deduplication log stores finite amount of `block_id`'s. That set of `block_id`'s determines how deduplication works.
```

File: docs/en/guides/developer/insertion-deduplication-on-retries.md [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+The profile setting `insert_deduplicate` controls deduplication on the query level. Note: all the data which is inserted with `insert_deduplicate=0` could not be deduplicated with the following insertion retry with `insert_deduplicate=1`. There is two reasons for that. First -- there are no `block_id`'s writted for the blocks from the insertion with the settions `insert_deduplicate=0`. Second -- user has to perform retries with the same settins as original operation.
```
Comment (kssenii) [inline docs/en/guides/developer/insertion-deduplication-on-retries.md:17]:
```suggestion
The profile setting `insert_deduplicate` controls deduplication on the query level. Note: all the data which is inserted with `insert_deduplicate=0` would not be deduplicated with the following insertion retry with `insert_deduplicate=1`. There is two reasons for that. First -- there are no `block_id`'s writted for the blocks from the insertion with the settions `insert_deduplicate=0`. Second -- user has to perform retries with the same settins as original operation.
```

File: docs/en/guides/developer/insertion-deduplication-on-retries.md [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+
+# How insertion deduplication works?
+When some data is inserted to the Clickhouse its is splited by rows count and bytes count to the sequence of the blocks. For each such block `*MergeTree` engine calculates hash from the data in that block. That hash is called `block_id` and it is used as a unique key for that operation. That approach works just well enough for the most cases assuming that different insertions contain different data. Othervise you need to use insertion setting `insert_deduplication_token`. That setting gives you precise control over deduplication process. Insertions only with equal `insert_deduplication_token` is deduplicated with each other.
```
Comment (kssenii) [inline docs/en/guides/developer/insertion-deduplication-on-retries.md:20]:
```suggestion
When some data is inserted to the Clickhouse, it is split by rows count and bytes count to the sequence of the blocks. For each such block `*MergeTree` engine calculates hash from the data in that block. That hash is called `block_id` and it is used as a unique key for that operation. That approach works just well enough for the most cases assuming that different insertions contain different data. Othervise you need to use insertion setting `insert_deduplication_token`. That setting gives you precise control over deduplication process. Insertions only with equal `insert_deduplication_token` is deduplicated with each other.
```

File: docs/en/guides/developer/insertion-deduplication-on-retries.md [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+
+Each time that data is inserted to the destination table, `block_id` is written to the deduplication log. For each insertion the condition is checked that there is no such `block_id` in deduplication log. If `block_id` is found in deduplication log, than the block is considered as duplicate. Note that Deduplication log stores finite count of `block_id`'s. Only insertions which meet the deduplication log window parameters have chances to be deduplicated.
```
Comment (kssenii) [inline docs/en/guides/developer/insertion-deduplication-on-retries.md:22]:
```suggestion
Each time that data is inserted to the destination table, `block_id` is written to the deduplication log. For each insertion the condition is checked that there is no such `block_id` in deduplication log. If `block_id` is found in deduplication log, then the block is considered as duplicate. Note that Deduplication log stores finite count of `block_id`'s. Only insertions which meet the deduplication log window parameters have chances to be deduplicated.
```

File: docs/en/guides/developer/insertion-deduplication-on-retries.md [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+
+For `INSERT VALUES` queries splitting the inserted data to the block is deterministic and it is determined by settings. Therefore user shoud retry insertions with the same settings values as they were at first operation.
```
Comment (kssenii) [inline docs/en/guides/developer/insertion-deduplication-on-retries.md:24]:
```suggestion
For `INSERT VALUES` queries splitting the inserted data to the block is deterministic and it is determined by settings. Therefore user should retry insertions with the same settings values as they were at first operation.
```

File: docs/en/guides/developer/insertion-deduplication-on-retries.md [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+
+For `INSERT SELECT` queries it is important that `SELECT` part of the query returns the same data in the same order each try. That is hard to achive in practical usage for many reasons. In order to achieve the stable data order on retries you could define precise `ORDER BY` section in `SELECT` part of the query. But the selected table could be updated between retries therefore the result data could change. Also there are could be a lot of data as a result a lot of blocks after sptitting inserted data by rows count and bytes count. That count of blocks might overflow deduplication log window.
```
Comment (kssenii) [inline docs/en/guides/developer/insertion-deduplication-on-retries.md:26]:
```suggestion
For `INSERT SELECT` queries it is important that `SELECT` part of the query returns the same data in the same order each try. That is hard to achieve in practical usage for many reasons. In order to achieve the stable data order on retries you could define precise `ORDER BY` section in `SELECT` part of the query. But the selected table could be updated between retries therefore the result data could change. Also there are could be a lot of data as a result a lot of blocks after splitting inserted data by rows count and bytes count. That count of blocks might overflow deduplication log window.
```

File: docs/en/guides/developer/insertion-deduplication-on-retries.md [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+
+
+# How insertion deduplication works with materialized views?
+When a table has one or several materialized views, the inserted data is also inserted to the destination of that views with some defined transformation.
```
Comment (kssenii) [inline docs/en/guides/developer/insertion-deduplication-on-retries.md:30]:
```suggestion
When a table has one or several materialized views, the inserted data is also inserted to the destination of those views with some defined transformation.
```

File: docs/en/guides/developer/insertion-deduplication-on-retries.md [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+That transformed data also has to be deduplicated on retries. Clickhouse deduplicates it in the same way it deduplicates data which is inserted into target table.
+User could control that process with settings on the table under materialized view: `replicated_deduplication_window`, `replicated_deduplication_window_seconds` and `non_replicated_deduplication_window`. Also user could use profile setting `deduplicate_blocks_in_dependent_materialized_views`.
+For the blocks inserted in tables under materialized views Clickhouse calulates `block_id` as hash from a string which contains concatenation of `block_id`'s for the the source table and other parts which helps to distiguish blocks afret materialised views transformation, like source view's id and the sequential number of that block. That makes deduplication for materialised views works correctly and distinguish the data by its original inserted data, no mater how it has been transormed on its way to the destination table under materialized view.
```
Comment (kssenii) [inline docs/en/guides/developer/insertion-deduplication-on-retries.md:33]:
```suggestion
For the blocks inserted in tables under materialized views Clickhouse calculates `block_id` as hash from a string which contains concatenation of `block_id`'s for the the source table and other parts which helps to distinguish blocks after materialised views transformation, like source view's id and the sequential number of that block. That makes deduplication for materialised views working correctly by distinguishing the data by its original inserted data, no matter how it has been transformed on its way to the destination table under materialized view.
```

File: docs/en/guides/developer/insertion-deduplication-on-retries.md [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+
```
Comment (kssenii) [inline docs/en/guides/developer/insertion-deduplication-on-retries.md:34]: Also mention that `update_insert_deduplication_token_in_dependent_materialized_views` is deprecated after some version?

**Review by justindeguzman — APPROVED** [review]

**Comment by CheSema** [comment]
Did this docs go to the production?  The feature has not been merged yet...

**Comment by justindeguzman** [comment]
Sorry about that, will revert asap. I'll just hide it from the sidebar and it won't be accessible. I thought you wanted me to edit it and then merge.

**Comment by CheSema** [comment]
Yes, I want that.
But I think it is better to merge it together with the feature.
https://github.com/ClickHouse/ClickHouse/pull/61601

I should have filled the description for that PR as well, sorry.

**Comment by justindeguzman** [comment]
Yes this was removed

**Comment by CheSema** [comment]
The code is committed, I resurrect this doc in a day or two.

**Pull request merged**